src/Storages/StorageFile.cpp

```diff
@@ -606,7 +606,7 @@ class StorageFileSource : public ISource
                     read_buf = createReadBuffer(current_path, storage->use_table_fd, storage->getName(), storage->table_fd, storage->compression_method, context);
 
                 auto format
-                    = context->getInputFormat(storage->format_name, *read_buf, block_for_format, max_block_size, storage->format_settings);
+                    = context->getInputFormat(storage->format_name, *read_buf, block_for_format, max_block_size, storage->format_settings, storage->paths.size());
```

Comment:
it seems that we should use here `num_streams` from here https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/StorageFile.cpp#L735 instead of `paths.size()`, shouldn't we? because `num_streams` is the actual of files being processed in parallel?

Reply:
Basically if there are many files to read, we want to lower the number of threads for parallel parsing from a single file. So I think we should do this here. In `read` function, we use `num_streams` to reserve Pipes and send all `filesInfo` to each `StorageFileSource`. In `generate` we go to each file, and also create `ParallelParsingInputFormat` (from `getInputFormat`).
https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/StorageFile.cpp#L591

Reply:
>  In generate we go to each file, and also create ParallelParsingInputFormat (from getInputFormat).

yes, but each file is not processed in parallel, only std::min(max_streams, paths.size()) files are processed in parallel so at a moment of time only this number of `ParallelParsingInputFormat` can be used, so we could use a higher value of `max_parsing_threads` than it would be if you divided by paths.size(). 
Though this is probably not important because as I remember usually max_streams will normally be equal to max_threads, so here it will not make much difference.